sql/catalyst/src/test/scala/org/apache/spark/sql/catalyst/optimizer/complexTypesSuite.scala

```diff
@@ -0,0 +1,499 @@
+      GetStructField(e, ord, Some(f))
+    }
+
+    def getArrayStructField(f : String) : Expression = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[ArrayType])
+      val arrType = e.dataType.asInstanceOf[ArrayType]
+      assert(arrType.elementType.isInstanceOf[StructType])
+      val structType = arrType.elementType.asInstanceOf[StructType]
+      val ord = structType.fieldNames.indexOf(f)
+      assert(-1 != ord)
+      GetArrayStructFields(e, structType(ord), ord, 1, arrType.containsNull)
+    }
+
+    def getArrayItem(i : Int) : GetArrayItem = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[ArrayType])
+      GetArrayItem(e, Literal(i))
+    }
+
+    def getMapValue(k : Expression) : Expression = {
+      assert(e.resolved)
+      assert(e.dataType.isInstanceOf[MapType])
+      val mapType = e.dataType.asInstanceOf[MapType]
+      assert(k.dataType == mapType.keyType)
+      GetMapValue(e, k)
+    }
+
+    def addCaseWhen( cond : Expression, v : Expression) : Expression = {
+      assert(e.resolved)
+      assert(e.isInstanceOf[CaseWhen])
+      assert(cond.dataType == BooleanType)
+      assert(v.dataType == e.dataType)
+      val CaseWhen(branches, default) = e
+      CaseWhen( branches :+ (cond, v), default)
+    }
+  }
+
+  test("explicit") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil).getStructField("att") as "outerAtt"
+   )
+
+    assertResult(StructType(StructField("outerAtt", LongType, nullable = false) :: Nil))(rel.schema)
+
+    val optimized = Optimize execute rel
+
+    val expected = baseOptimizedPlan.select(idRef as "outerAtt")
+
+    comparePlans(optimized, expected)
+  }
+
+  ignore("explicit - deduced att name") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil).getStructField("att")
+   )
+    assertResult(StructType(
+      StructField("named_struct(att, id AS `att`).att", LongType, nullable = false) :: Nil
+    ))(rel.schema)
+
+    val optimized = Optimize execute rel
+
+    val expected = baseOptimizedPlan.select(idRef as "named_struct(att, id AS `att`).att")
+
+    comparePlans(optimized, expected)
+  }
+
+  test("collapsed") {
+    val rel = baseOptimizedPlan.select(
+      CreateNamedStruct("att" :: idRef :: Nil) as "struct1"
+    )
+    assertResult(
+      StructType(
+        StructField(
+          "struct1",
+          StructType(StructField("att", LongType, false) :: Nil),
+          false
+       ) :: Nil
+      )
+    )(rel.schema)
+
+    val struct1Ref = rel.output.head
+    val rel2 = rel.select(struct1Ref.getStructField("att").as("struct1Att"))
+
+    assertResult(
+      StructType(
+        StructField("struct1Att", LongType, false) :: Nil
+     )
+    )(rel2.schema)
+
+    val optimized = Optimize execute rel2
+    val expected =
+      baseOptimizedPlan.select(idRef as "struct1Att" )
+
+    comparePlans(optimized, expected)
+  }
+
+  test("collapsed2") {

```

Comment:
Please make the test names more descriptive. For example this one could be named: `collapse multiple CreateNamedStruct/GetStructField pairs`